File: backend/LimsAuth.Api/Models/Schedule/UsageRegistration.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace LimsAuth.Api.Models;

[Table("usage_registrations")]
public class UsageRegistration
{
    [Key]
    [Column("id")]
    public Guid Id { get; set; } = Guid.NewGuid();

    [Required]
    [Column("semester_id")]
    public Guid SemesterId { get; set; }

    [Column("lab_id")]
    public Guid? LabId { get; set; }

    [Required]
    [Column("lab_name")]
    [MaxLength(100)]
    public string LabName { get; set; } = string.Empty;

    [Column("building_name")]
    [MaxLength(100)]
    public string? BuildingName { get; set; }

    [Column("room_number")]
    [MaxLength(50)]
    public string? RoomNumber { get; set; }

    [Column("use_date")]
    public DateTime UseDate { get; set; }

    [Column("week_number")]
    public int WeekNumber { get; set; }

    [Column("day_of_week")]
    public int DayOfWeek { get; set; }

    [Column("period_number")]
    public int PeriodNumber { get; set; }

    [Column("source")]
    [MaxLength(50)]
    public ScheduleSource Source { get; set; }

    [Column("schedule_entry_id")]
    public Guid? ScheduleEntryId { get; set; }

    [Column("reservation_id")]
    public Guid? ReservationId { get; set; }

    [Column("teaching_application_id")]
    public Guid? TeachingApplicationId { get; set; }

    [Column("experiment_task_id")]
    public Guid? ExperimentTaskId { get; set; }

    [Column("teaching_task_id")]
    public Guid? TeachingTaskId { get; set; }

    [Column("course_name")]
    [MaxLength(200)]
    public string? CourseName { get; set; }

    [Column("project_name")]
    [MaxLength(200)]
    public string? ProjectName { get; set; }

    [Column("experiment_item_name")]
    [MaxLength(200)]
    public string? ExperimentItemName { get; set; }

    [Column("experiment_item_type")]
    [MaxLength(50)]
    public string? ExperimentItemType { get; set; }

    [Column("planned_hours")]
    public int PlannedHours { get; set; }

    [Column("actual_hours")]
    public double ActualHours { get; set; }

    [Column("class_name")]
    [MaxLength(100)]
    public string? ClassName { get; set; }

    [Column("expected_student_count")]
    public int? ExpectedStudentCount { get; set; }

    [Column("actual_student_count")]
    public int? ActualStudentCount { get; set; }

    [Column("attendance_record")]
    [MaxLength(500)]
    public string? AttendanceRecord { get; set; }

    [Column("teaching_condition")]
    [MaxLength(100)]
    public string? TeachingCondition { get; set; }

    [Column("equipment_condition")]
    [MaxLength(100)]
    public string? EquipmentCondition { get; set; }

    [Column("status")]
    [MaxLength(20)]
    public RegistrationStatus Status { get; set; } = RegistrationStatus.Pending;

    [Column("reminded_at")]
    public DateTime? RemindedAt { get; set; }

    [Required]
    [Column("filled_by_id")]
    public Guid FilledById { get; set; }

    [Required]
    [Column("filled_by_name")]
    [MaxLength(100)]
    public string FilledByName { get; set; } = string.Empty;

    [Column("filled_at")]
    public DateTime FilledAt { get; set; } = DateTime.UtcNow;

    [Column("created_at")]
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    [Column("created_by")]
    public string? CreatedBy { get; set; }

    [Column("updated_at")]
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;

    [Column("updated_by")]
    public string? UpdatedBy { get; set; }

    public virtual Semester? Semester { get; set; }
    public virtual Lab? Lab { get; set; }
    public virtual User? FilledBy { get; set; }
}

File: mobile/lib/models/usage_registration.dart
import 'lab.dart';
import '../utils/num_utils.dart';

class UsageRegistration {
  final String id;
  final String? semesterId;
  final String? semesterName;
  final String? labId;
  final String? labName;
  final String? useDate;
  final int weekNumber;
  final int dayOfWeek;
  final int periodNumber;
  final String? source;
  final String? scheduleEntryId;
  final String? reservationId;
  final String? teachingApplicationId;
  final String? courseName;
  final String? projectName;
  final String? experimentItemName;
  final String? experimentItemType;
  final int plannedHours;
  final double actualHours;
  final String? className;
  final int expectedStudentCount;
  final int actualStudentCount;
  final String? attendanceRecord;
  final String? teachingCondition;
  final String? equipmentCondition;
  final String? status;
  final String? remindedAt;
  final String? filledById;
  final String? filledByName;
  final String? filledAt;
  final String? createdAt;
  final Lab? lab;

  UsageRegistration({
    required this.id,
    this.semesterId,
    this.semesterName,
    this.labId,
    this.labName,
    this.useDate,
    this.weekNumber = 0,
    this.dayOfWeek = 0,
    this.periodNumber = 0,
    this.source,
    this.scheduleEntryId,
    this.reservationId,
    this.teachingApplicationId,
    this.courseName,
    this.projectName,
    this.experimentItemName,
    this.experimentItemType,
    this.plannedHours = 0,
    this.actualHours = 0,
    this.className,
    this.expectedStudentCount = 0,
    this.actualStudentCount = 0,
    this.attendanceRecord,
    this.teachingCondition,
    this.equipmentCondition,
    this.status,
    this.remindedAt,
    this.filledById,
    this.filledByName,
    this.filledAt,
    this.createdAt,
    this.lab,
  });

  factory UsageRegistration.fromJson(Map<String, dynamic> json) {
    return UsageRegistration(
      id: json['id']?.toString() ?? '',
      semesterId: json['semesterId']?.toString(),
      semesterName: json['semesterName']?.toString(),
      labId: json['labId']?.toString(),
      labName: json['labName']?.toString(),
      useDate: json['useDate']?.toString(),
      weekNumber: safeInt(json['weekNumber']),
      dayOfWeek: safeInt(json['dayOfWeek']),
      periodNumber: safeInt(json['periodNumber']),
      source: json['source']?.toString(),
      scheduleEntryId: json['scheduleEntryId']?.toString(),
      reservationId: json['reservationId']?.toString(),
      teachingApplicationId: json['teachingApplicationId']?.toString(),
      courseName: json['courseName']?.toString(),
      projectName: json['projectName']?.toString(),
      experimentItemName: json['experimentItemName']?.toString(),
      experimentItemType: json['experimentItemType']?.toString(),
      plannedHours: safeInt(json['plannedHours']),
      actualHours: safeDouble(json['actualHours']),
      className: json['className']?.toString(),
      expectedStudentCount: safeInt(json['expectedStudentCount']),
      actualStudentCount: safeInt(json['actualStudentCount']),
      attendanceRecord: json['attendanceRecord']?.toString(),
      teachingCondition: json['teachingCondition']?.toString(),
      equipmentCondition: json['equipmentCondition']?.toString(),
      status: json['status']?.toString(),
      remindedAt: json['remindedAt']?.toString(),
      filledById: json['filledById']?.toString(),
      filledByName: json['filledByName']?.toString(),
      filledAt: json['filledAt']?.toString(),
      createdAt: json['createdAt']?.toString(),
      lab: json['lab'] != null
          ? Lab.fromJson(json['lab'] as Map<String, dynamic>)
          : null,
    );
  }

  Map<String, dynamic> toJson() => {
        'semesterId': semesterId,
        'labId': labId,
        'labName': labName,
        'useDate': useDate,
        'weekNumber': weekNumber,
        'dayOfWeek': dayOfWeek,
        'periodNumber': periodNumber,
        'source': source,
        'scheduleEntryId': scheduleEntryId,
        'reservationId': reservationId,
        'teachingApplicationId': teachingApplicationId,
        'courseName': courseName,
        'projectName': projectName,
        'experimentItemName': experimentItemName,
        'experimentItemType': experimentItemType,
        'plannedHours': plannedHours,
        'actualHours': actualHours,
        'className': className,
        'expectedStudentCount': expectedStudentCount,
        'actualStudentCount': actualStudentCount,
        'attendanceRecord': attendanceRecord,
        'teachingCondition': teachingCondition,
        'equipmentCondition': equipmentCondition,
      };
}

class CompletionRate {
  final int total;
  final int completed;
  final int pending;
  final int overdue;
  final double rate;

  CompletionRate({
    this.total = 0,
    this.completed = 0,
    this.pending = 0,
    this.overdue = 0,
    this.rate = 0,
  });

  factory CompletionRate.fromJson(Map<String, dynamic> json) {
    return CompletionRate(
      total: safeInt(json['total']),
      completed: safeInt(json['completed']),
      pending: safeInt(json['pending']),
      overdue: safeInt(json['overdue']),
      rate: safeDouble(json['rate']),
    );
  }
}
